airgun/widgets.py

```diff
@@ -267,3 +274,136 @@ def fill(self, values):
+    confirm_dialog = Text(".//button[contains(@ng-click, 'ok')]")
+    cancel_dialog = Text(
+        ".//button[contains(@ng-click, 'cancel') and contains(@class, 'btn')]")
+    discard_dialog = Text(
+        ".//button[contains(@ng-click, 'cancel') and @class='close']")
+
+    def confirm(self):
+        """Clicks on the positive outcome button like 'Remove', 'Ok', 'Yes'"""
+        self.confirm_dialog.click()
+
+    def cancel(self):
+        """Clicks on the negative outcome button like 'Cancel' or 'No'"""
+        self.cancel_dialog.click()
+
+    def read(self):
+        do_not_read_this_widget()
+
+
+class CheckboxGroup(Widget):
+    """Group of checkboxes that goes in a line one after another. Usually used
+    to specify lifecycle environment
+
+    Example html representation::
+
+        <//ul[@class='path-list']>
+            <li class="path-list-item ng-scope"...>
+                <label class="path-list-item-label...>
+                    <input type="checkbox"...>
+            <li class="path-list-item ng-scope"...>
+                <label class="path-list-item-label...>
+                    <input type="checkbox"...>
+
+    Locator example::
+
+        //ul[@class='path-list']
+
+    """
+    ROOT = "//ul[@class='path-list']"
+    LABELS = "./li/label[contains(@class, path-list-item-label)]"
+    CHECKBOX = (
+        ".//input[@ng-model='item.selected'][parent::label[contains(., '{}')]]"
+    )
+
+    def cb_selected(self, locator):
+        """Identify whether specific checkbox is selected or not"""
+        return 'ng-not-empty' in self.browser.get_attribute('class', locator)
+
+    def select(self, locator, value):
+        """Select or deselect checkbox depends on the value passed"""
+        value = bool(value)
+        current_value = self.cb_selected(locator)
+        if value == current_value:
+            return False
+        else:
+            self.browser.element(locator).click()
+            if self.cb_selected(locator) != value:
+                raise WidgetOperationFailed(
+                    'Failed to set the checkbox to requested value.')
+            return True
+
+    def read(self):
+        """Return a list of dictionaries. Each dictionary consists of name and
+        value for each checkbox from the group
+        """
+        checkboxes = []
+        for item in self.browser.elements(self.LABELS):
+            name = self.browser.text(item)
+            value = self.cb_selected(self.CHECKBOX.format(name))
+            checkboxes.append({'name': name, 'value': value})
+        return checkboxes
+
+    def fill(self, values):
+        """Assign value for specific checkbox from group
+
+        :param values: dictionary that consist of checkbox name and value
+            that should be assigned to that checkbox
+        """
+        checkbox_locator = self.CHECKBOX.format(values.get('entity_name'))
+        return self.select(checkbox_locator, values.get('select'))
```

Comment:
Helpers like `select` and `cb_selected` accept locators and create web elements multiple time. Are you sure that's the optimal way? I'd rather form web element for checkbox here and passed that single element to all the helpers.

Reply:
You can't do that as `fill()` and `read()` has no connection with each other